planning/behavior_velocity_planner/autoware_behavior_velocity_blind_spot_module/src/time_to_collision.cpp

```diff
@@ -0,0 +1,239 @@
+    path.points, current_pose, nearest_dist_threshold, nearest_yaw_threshold);
+  double passing_time = time_to_restart;
+
+  std::vector<FuturePosition> future_positions;
+  future_positions.emplace_back(FuturePosition{path.points.at(closest_idx), passing_time});
+  for (unsigned i = closest_idx + 1; i + 1 < path.points.size(); ++i) {
+    const auto & p1 = path.points.at(i);
+    const auto & p2 = path.points.at(i + 1);
+    const double dist = autoware_utils_geometry::calc_distance2d(p1, p2);
+    const double average_velocity =
+      (p1.point.longitudinal_velocity_mps + p2.point.longitudinal_velocity_mps) / 2.0;
+    const double passing_velocity = std::max(average_velocity, 0.01);
+    passing_time += dist / passing_velocity;
+
+    future_positions.emplace_back(FuturePosition{path.points.at(i), passing_time});
+  }
+  return future_positions;
+}
+
+std::vector<FuturePosition> calculate_future_profile(
+  const autoware_internal_planning_msgs::msg::PathWithLaneId & path, const double time_to_restart,
+  const std::shared_ptr<const PlannerData> & planner_data, const lanelet::Id lane_id)
+{
+  const double nearest_dist_threshold = planner_data->ego_nearest_dist_threshold;
+  const double nearest_yaw_threshold = planner_data->ego_nearest_yaw_threshold;
+  const auto & current_pose = planner_data->current_odometry->pose;
+  const double current_velocity = planner_data->current_velocity->twist.linear.x;
+
+  const auto closest_idx = autoware::motion_utils::findFirstNearestIndexWithSoftConstraints(
+    path.points, current_pose, nearest_dist_threshold, nearest_yaw_threshold);
+
+  autoware_internal_planning_msgs::msg::PathWithLaneId reference_path;
+  bool assigned_lane_found = false;
+  for (unsigned i = 0; i < path.points.size(); ++i) {
+    auto reference_point = path.points.at(i);
+    // assume backward velocity is current ego velocity
+    if (i < closest_idx) {
+      reference_point.point.longitudinal_velocity_mps = static_cast<float>(current_velocity);
+    }
+    reference_path.points.push_back(reference_point);
+    const bool has_objective_lane_id = has_lane_ids(path.points.at(i), lane_id);
+    if (assigned_lane_found && !has_objective_lane_id) {
+      break;
+    }
+    assigned_lane_found = has_objective_lane_id;
+  }
+  if (reference_path.points.size() < 3 || !assigned_lane_found) {
+    return {};
+  }
+  auto smoothed_reference_path = reference_path;
+  if (!smoothPath(reference_path, smoothed_reference_path, planner_data)) {
+    return {};
+  }
+  return calculate_future_profile_impl(
+    smoothed_reference_path, current_pose, time_to_restart, nearest_dist_threshold,
+    nearest_yaw_threshold);
+}
+
+std::optional<std::pair<double, double>> compute_time_interval_for_passing_line(
+  const std::vector<FuturePosition> & future_positions,
+  const autoware_utils_geometry::LinearRing2d & footprint, const lanelet::ConstLineString3d & line1,
+  const lanelet::ConstLineString3d & line2)
```

Comment:
I think some additional filtering can be performed using the `ego_passage_interval` in this function.
If I understand correctly, we only care about the position of the object during `ego_passage_interval` (+/- the TTC margins). So I think the calculations can be skipped when the time is too far from the `ego_passage_interval`.

This should improve performance and may also fix the current bug with the false positives.